### ISSUE TITLE
[BUG][Example][wrong-code] `_tir_packed_to_unsigned_convert_with_zeros` computes `q - zero` in the unsigned storage dtype, silently wrapping instead of dequantizing

### ISSUE THREAD

**Issue by Hughshine** [tile-ai/tilelang#2947]
### Required prerequisites

- [x] I have read the documentation <https://tilelang.com>.
- [x] I have searched the [Issue Tracker](https://github.com/tile-ai/tilelang/issues) and [Discussions](https://github.com/tile-ai/tilelang/discussions) that this hasn't already been reported. (+1 or comment there if it has.)

### What version of TileLang are you using?

0.1.13 (latest release). On 0.1.13 the helper lives under `examples/dequantize_gemm/quantize/quantization.py` (relocated from `tilelang/quantize/`); the buggy line is unchanged.

### System information

TileLang 0.1.13 / CUDA 12.8 / PyTorch 2.8.0, NVIDIA L40S (sm_89).

### Problem description

`_tir_packed_to_unsigned_convert_with_zeros` (the zero-point dequant helper; `examples/dequantize_gemm/quantize/quantization.py` on 0.1.13, formerly `tilelang.quantize`) subtracts the zero-point **before** the cast to the output dtype and **in the unsigned storage dtype**, so whenever a quantized value is smaller than its zero-point — the normal case for asymmetric quantization — the subtraction wraps modulo `2**storage_nbit` and the decoded value is silently wrong.

With `uint32` packed storage, decoding nibbles `0..15` with `zero=8` and a `float16` output produces `inf` for every field where `q < zero`, instead of `-8.0, -7.0, …`:

```
got  [inf, inf, inf, inf, inf, inf]
exp  [-8.0, -7.0, -6.0, -5.0, -4.0, -3.0]        # signed (q - zero)
# same helper with zero=0 (no field underflows) decodes correctly
```

The emitted CUDA is `(half_t)(((P[i] >> 0) & 15u) - 8u)`: `0u - 8u` evaluates to `4294967288u` in `uint32`, and `(half_t)4294967288u` is `inf`. No error is raised; the kernel compiles and runs to completion. Not a regression — see Provenance.

<details>
<summary>Output dtype and storage width only change the wrong value, not the defect</summary>

The defect is the unsigned subtract; the symptom depends on the output dtype (measured, `uint32` storage, nibble `0`, `zero=8`, expected `-8`): `float16` gives `inf` (the wrapped `uint32` exceeds the fp16 range), `float32` gives `4.29e9` (≈`2**32`). An `int32` output happens to come back correct (`-8`), because the `uint32→int32` reinterpret of the wrapped bit pattern `0xFFFFFFF8` is exactly `-8` — so integer outputs mask the bug the same way narrow storage does, while float outputs expose it.

The `uint8`/`uint16` storage variants happen to return the right answer for small fields, but only by C accident: `uchar - uchar` is promoted to `int` in the generated CUDA before the cast, so the wrap doesn't occur. The `uint32` (and `uint64`) storage widths get no such promotion. So `storage_nbit=32` reliably reproduces; `storage_nbit=8` masks the same source-level bug.

</details>

### Reproducible example code

```python
import sys, numpy as np, torch, tilelang, tilelang.language as T
from tvm import tirx                                    # 0.1.13: tvm.tir -> tvm.tirx

# 0.1.13: the helper moved from `tilelang.quantize` into the dequantize-gemm
# example, which is NOT shipped in the pip wheel — point at a source checkout's
# examples/dequantize_gemm/quantize dir and import the helper from there.
sys.path.insert(0, "/path/to/tilelang/examples/dequantize_gemm/quantize")
from quantization import _tir_packed_to_unsigned_convert_with_zeros

N, ZERO = 16, 8                                        # asymmetric zero-point
fz = _tir_packed_to_unsigned_convert_with_zeros("uint", 32)   # uint32 packed storage

def build(zero):
    @T.prim_func
    def deq(P: T.Tensor((N,), "uint32"), O: T.Tensor((N,), "float16")):
        with T.Kernel(1, threads=N) as bx:
            t = T.get_thread_binding()
            O[t] = fz(4, P[t], tirx.const(0, "int32"), tirx.const(zero, "uint32"), "float16")
    return deq

q = np.arange(16, dtype=np.uint32)                     # quantized nibbles 0..15

# TRIGGER: zero=8, so q<8 underflows in uint32
kt  = tilelang.compile(build(ZERO), out_idx=[1])
got = kt(torch.tensor(q, device="cuda")).cpu().numpy().astype(np.float64)
ref = np.array([float(int(v) - ZERO) for v in q])      # signed (q - zero)
print("got", got[:6].tolist())                         # -> [inf, inf, inf, inf, inf, inf]
print("exp", ref[:6].tolist())                         # -> [-8.0, -7.0, -6.0, -5.0, -4.0, -3.0]
assert np.max(np.abs(got - ref)) < 1e-3                # -> AssertionError (max_err = inf)

# CONTROL: zero=0, no field underflows -> correct
kc   = tilelang.compile(build(0), out_idx=[1])
gotc = kc(torch.tensor(q, device="cuda")).cpu().numpy().astype(np.float64)
assert np.max(np.abs(gotc - q.astype(np.float64))) < 1e-3   # -> passes
```

### Traceback

No traceback — the kernel compiles and runs to completion; the decoded values are silently wrong (`inf` / wrapped) and deterministic.

### Expected behavior

`q - zero` should be evaluated in the signed/output domain so it can be negative, matching the sibling `_tir_packed_to_signed_convert`, which casts to the output dtype **first** and subtracts a constant **after** ([quantization.py#L305](https://github.com/tile-ai/tilelang/blob/8001cc4ccf6149382d2019654a19f59c1d4d0482/examples/dequantize_gemm/quantize/quantization.py#L305)): `(...).astype(dtype) - tirx.const(max_int_value, dtype)`. Decoding nibbles `0..15` with `zero=8` is expected to give `-8.0 … 7.0`.

### Additional context

**Root cause.** The zero-point subtract is performed in the unsigned storage dtype before the widening cast, so an underflow (`q < zero`) wraps modulo the storage width instead of producing a negative number. The whole point of a zero-point is that `q - zero` is signed, so this is the normal input range for the helper.

The offending line ([quantization.py#L328](https://github.com/tile-ai/tilelang/blob/8001cc4ccf6149382d2019654a19f59c1d4d0482/examples/dequantize_gemm/quantize/quantization.py#L328)):

```python
return (((val >> (pos * nbit).astype(storage_dtype)) & mask) - zero).astype(dtype)
#        ^--------- unsigned storage_dtype -----------------------^  ^ subtract here, then cast
```

**Suggested fix.** Move the cast inside the subtraction so the difference is computed in the output domain, matching the signed sibling `_tir_packed_to_signed_convert`:

```python
return ((val >> (pos * nbit).astype(storage_dtype)) & mask).astype(dtype) - zero.astype(dtype)
```

Whether `zero` is cast to `dtype` or to a signed integer first depends on the intended `zero` dtype (not verified end-to-end).

**Provenance.** Introduced with the helper in [#494](https://github.com/tile-ai/tilelang/pull/494) (`cde1886f2`, 2025-05-14, "Introduce quantize components of TileLang"); the `- zero).astype(dtype)` form is unchanged since. A related sign-extension fix landed for the *signed* uint32 decode in [#2500](https://github.com/tile-ai/tilelang/pull/2500), but the `with_zeros` path was not touched. The file was relocated into `examples/dequantize_gemm/quantize/` by [#2761](https://github.com/tile-ai/tilelang/pull/2761); the helper and the buggy line are byte-identical there on `main`. The code is adapted from mlc-llm's `quantization.py` per the file header.

**Dedup.** I searched the open and closed tracker and found no existing report of this defect.

**Reach.** `_tir_packed_to_unsigned_convert_with_zeros` is an internal, underscore-prefixed helper whose sole purpose is asymmetric zero-point dequant, where `q - zero` is meant to be signed. Note on scope: since #2761 it lives under `examples/dequantize_gemm/quantize/` and is only re-exported from that example package's own `__init__.py` — it is **not** part of the installed `tilelang` wheel (there is no `tilelang/quantize/` in the pip package on 0.1.13), so it is reached by building the dequantize-GEMM example from a source checkout, not by importing a public `tilelang` API. Before #2761 it lived in `tilelang/quantize/` and was shipped. `grep` of the repo: the only sites naming it are its definition and its re-export in `__init__.py` — no example or test in `examples/`/`testing/` calls it with a non-zero `zero`, which is why CI is green. The `storage_nbit=8` default masks the bug through C integer promotion; the `uint32`/`uint64` storage widths (used for 32-bit-packed weight layouts) hit it whenever any quantized field is below its zero-point.

**Impact.** The trigger is narrow — it needs 32-/64-bit packed storage plus a non-zero zero-point where at least one quantized field falls below it. But when it fires it is a silent wrong-code miscompile: no error is raised, and every field with `q < zero` decodes to a wrapped value (`inf`, a large positive float, or a wrapped integer) across the whole dequantized weight region rather than one bounded value. Because whether it fires is data-dependent (only fields below their zero-point), it can pass small tests and slip into an asymmetric-quantization workload unnoticed. Fixing it closes the unsigned-underflow boundary and aligns the `with_zeros` path with the signed sibling's cast-then-subtract order.
